curve25519-dalek/src/lemmas/field_lemmas/field_algebra_lemmas.rs

```diff
@@ -0,0 +1,999 @@
+        lemma_mul_mod_noop((x * y) as int, (x * y) as int, p as int);
+
+        // Step 2: Show (x*x)*(y*y) = (x*y)*(x*y) as integers
+        // (x*y)*(x*y) = x*(y*x)*y = x*(x*y)*y = (x*x)*(y*y)
+        assert((x * y) * (x * y) == (x * x) * (y * y)) by {
+            // (x*y)*(x*y) = x * (y * (x * y))
+            lemma_mul_is_associative(x as int, y as int, (x * y) as int);
+            // y * (x * y) = (y * x) * y
+            lemma_mul_is_associative(y as int, x as int, y as int);
+            // y * x = x * y
+            lemma_mul_is_commutative(y as int, x as int);
+            // So x * (y * (x * y)) = x * ((x * y) * y) = x * (x * (y * y))
+            lemma_mul_is_associative(x as int, y as int, y as int);
+            // x * (x * (y * y)) = (x * x) * (y * y)
+            lemma_mul_is_associative(x as int, x as int, (y * y) as int);
+        };
+    };
+}
+
+// =============================================================================
+// Double Inverse and Solving Equations Lemmas
+// =============================================================================
+/// Lemma: inv(inv(x)) = x (mod p)
+///
+/// ## Mathematical Proof
+/// ```text
+/// Let y = inv(x). Then x · y = 1.
+/// Rearranging: y · x = 1.
+/// So x is the inverse of y = inv(x).
+/// By uniqueness: x = inv(inv(x)).
+/// ```
+pub proof fn lemma_inv_of_inv(x: nat)
+    requires
+        x % p() != 0,
+    ensures
+        math_field_inv(math_field_inv(x)) == x % p(),
+{
+    let p = p();
+    p_gt_2();  // Needed for field operations
+
+    let inv_x = math_field_inv(x);
+    let x_mod = x % p;
+
+    // Step 1: Get properties of inv(x): inv_x < p and (x % p) * inv_x % p == 1
+    assert(inv_x < p && ((x % p) * inv_x) % p == 1) by {
+        field_inv_property(x);
+    };
+
+    // Step 2: inv_x % p != 0 (otherwise (x % p) * 0 = 0 ≠ 1)
+    assert(inv_x % p != 0) by {
+        lemma_small_mod(inv_x, p);
+        if inv_x == 0 {
+            lemma_small_mod(0nat, p);
+            assert(false);  // 0 ≠ 1
+        }
+    };
+
+    // Step 3: x_mod < p
+    assert(x_mod < p) by {
+        lemma_mod_bound(x as int, p as int);
+    };
+
+    // Step 4: ((inv_x % p) * x_mod) % p == 1 (by commutativity)
+    assert(((inv_x % p) * x_mod) % p == 1) by {
+        lemma_small_mod(inv_x, p);
+        lemma_mul_is_commutative(inv_x as int, x_mod as int);
+    };
+
+    // Step 5: By uniqueness of inverse, x_mod = inv(inv_x)
+    assert(math_field_inv(inv_x) == x_mod) by {
+        field_inv_unique(inv_x, x_mod);
+    };
+}
+
+/// Lemma: If a · b = c (in field), then a = c · inv(b)
+///
+/// Useful for solving for one factor in a field equation.
+///
+/// ## Mathematical Proof
+/// ```text
+/// Given: a · b ≡ c (mod p)
+/// Multiply both sides by inv(b):
+///   a · b · inv(b) ≡ c · inv(b) (mod p)
+///   a · 1 ≡ c · inv(b) (mod p)
+///   a ≡ c · inv(b) (mod p)
+/// ```
+pub proof fn lemma_solve_for_left_factor(a: nat, b: nat, c: nat)
+    requires
+        b % p() != 0,
```

Comment:
I think this is also technically not required, because we define inv(0) = 0, but this is more of a remark than a request for removal. Maybe add a comment